### PR TITLE
CP-41573 & CP-41923: Add EUA check to RPU Wizard and bump `RPU005` to 5.0

### PR DESCRIPTION
**Pull request by danilo-delbusso**
We only expect Post82X hosts to have this capability. We will be adding a blocking precheck to ensure that all hosts can fetch the EUA. For testing, just ensure you're using the latest XenServer version. Feel free to ping me for help on how to test this locally. Please read the text below to get an understanding of why there are so many changes to `Check`s.

__________________________________

As part of this PR, I had to introduce the concept of a "permanent" check.

> Why can't we just add a normal check?

`Check` instances are created when the prechecks page is created or when the user hits on _Check again_ or _Resolve all_. This is usually fine because the `Check` objects look for changes in the properties or configuration of its associated `IXenObject`s to see if the `Problem` related to the `Check` has been resolved or not; e.g.: if a VM was running, it is simple to check that the VM is in the correct power state after running the action that resolves the problem.

However, we cannot use this mechanism to check if the user has accepted the EUA or not. We could technically update the pool's `gui-config` property, but XenCenter is not required to change properties of the pool when the user accepts the EUA and changing this property would introduce more issues.

> How does this PR address this issue?

We needed a way to check if the `Problem` had been solved by using the application's memory. To achieve this, some `Check` objects had to be kept across page reloads etc.. These are the _permanent checks_. They are saved in memory and use the new `check.Completed` property to check if their `Problem` has been resolved. They rely on the `Problem` instance to update the `Check`'s status.

To achieve this, `Check` and `Problem` objects related to permanent checks must override `Equals` and `GetHashCode` if they have custom properties (e.g. the `_targetUri` for the EUA `Check` and associated `Problem`). This is because the current logic only checks for changes in the `IXenObject`s related to the `Check`. If one were to not override `Equals`/`GetHashCode`, changes to the `_targetUri` property would not be kept, because XenCenter cannot know that the permanent check needs to be re-generated.

Hopefully we won't need to add many more of these special checks, but just in case I have added docs, so those + the usage in `UpgradeRequiresEua.cs` and `EuaNotAcceptedProblem` should help devs add similar `Check`s.

### CODE REVIEW

**Review by CitrixChris — APPROVED** [review]

**Review by kc284 — CHANGES_REQUESTED** [review]

File: XenAdmin/Wizards/PatchingWizard/PatchingWizard_PrecheckPage.cs [hunk 1]
```diff
@@ -429,6 +437,29 @@ protected virtual List<CheckGroup> GenerateChecks(Pool_patch patch)
             return groups;
         }
 
+
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Wizards/PatchingWizard/PatchingWizard_PrecheckPage.cs:441]: Visual Studio does render them though, that's why I found them useful to keep (hovering over a method will generate clickable links to the definitions mentioned in `cref`s)

[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Wizards/PatchingWizard/PatchingWizard_PrecheckPage.cs:441]: Alright, feel free to ignore the comment 😃

File: XenAdmin/Dialogs/AcceptEuaDialog.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            ThreadPool.QueueUserWorkItem(LoadEuas, null);
+        }
+
+        private void LoadEuas(object _)
+        {
+           _hosts.AsParallel().ForAll(FetchHostEua);
+
+            Program.BeginInvoke(this, () =>
+            {
+                if (_euas.Count > 1)
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Dialogs/AcceptEuaDialog.cs:71]: We don't need to and they should all show the same. But the idea is that I'm adding a fairly small amount of logic to support this in the future. While if we don't, we might forget that XenCenter only supports one EUA per RPU. I think in cases like these the extra complexity is warranted since it does prevent issues in case things change.

File: XenModel/Utils/Helpers.cs [hunk 3]
```diff
@@ -1580,5 +1581,33 @@ public static string AddAuthTokenToQueryString(string authToken, string existing
...
+                { "url", targetUri.ToString()}
+            };
+
+            try
+            {
+                var result = Host.call_plugin(host.Connection.Session, host.opaque_ref, "prepare_host_upgrade.py", "getEUA", args);
+                var jsonPayloadDefinition = new { eua = string.Empty };
+                var parsedPayload = JsonConvert.DeserializeAnonymousType(result, jsonPayloadDefinition);
+
+                eua = parsedPayload.eua;
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenModel/Utils/Helpers.cs:1603]: idea was for it to be caught and logged in case it happened (since it shouldn't)

File: XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    internal class UpgradeRequiresEua : UpgradeCheck
+    {
+        private readonly Uri _targetUri;
+        public UpgradeRequiresEua(List<Host> hosts, Uri targetUri)
+            : base(hosts)
+        {
+            _targetUri = targetUri;
+        }
+
+        public override bool CanRun() => Hosts.Any(Helpers.Post82X);
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs:50]: > This means the check is dependent on installing the RPU hotfix first (see for example SafeToUpgradeCheck).

For this, I was planning to add the relevant bits once we had an RPU (so within the RPU PR)

> If there is an error while retrieving the EUA though, I don't think we shoud proceed suddenly.

Unfortunately, the `getEUA` function returns `{ "eua" : null }` both in case of error and in case there's no EUA at all. We can circumvent this because we know that if the RPU is for Nile or greater the user needs to accept an EUA, but that is logic we'll need to add to XenCenter as opposed to relying on the `getEUA` function.

Not ideal in my opinion, since when/if things change we'll need to update both (e.g.: no EUA for a super special release? One never knows what's going to happen with these things 😅).

So to recap if XenCenter cannot fetch an EUA but we except one, we can add special wording instead of showing nothing. I was thinking something on the lines of `XenCenter could not fetch an End User Agreement (EUA), but expected one. Please go to https://example.com and check if your installation requires accepting an EUA. Selecting "I accept" means you have read and accept the terms of the EUA, if there is one.`

File: XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs [hunk 5]
```diff
@@ -294,6 +297,27 @@ where check.CanRun()
...
+                try
+                {
+                    targetUri = new Uri(uriText);
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex);
+                }
+
+                var euaCheck = GetPermanentCheck("EUA", new UpgradeRequiresEua(hostsToUpgrade, targetUri));
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs:314]: Not sure I agree with this. Moving the filtering to the caller means that whenever `UpgradeRequiresEua` is called, the developer needs to know and remember to filter the list before passing it to it. I think it's best if the `Check` itself knows that it only applies to post82x hosts. NowI know we will probably only ever have this `Check` called within this method, but it's just safer this way. Thoughts?

File: XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs:50]: I just noticed we already do something like this in the `AcceptEuaDialog`, where I show an error if the host is POST82X and we can't fetch the EUA

File: XenAdmin/Dialogs/AcceptEuaDialog.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Dialogs/AcceptEuaDialog.cs:71]: The problem with this though is that if the customer upgrades 5 pools, they will see a long text containing 5 times the same content, which is a bug from the user's perspective.

File: XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs:50]: Yes, something like the text you suggest above is exactly what I had in mind, I didn't imply adding the EUA within XC.
The error currently shown on AcceptEuaDialog does not cover us.

File: XenAdmin/Dialogs/AcceptEuaDialog.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Dialogs/AcceptEuaDialog.cs:71]: I see what you mean but they're stored in a `HashSet` so it's only going to show more than one if they're different.

File: XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs:314]: Moving the filtering to the caller may be needed in order to hide the check completely from the user if their system is not relevant. For example, running a check on havana for a feature that was added on naples will not cause an issue since the host will be filtered out, but the anachronism may confuse the user. At least that was the thinking - I'm not sure we get it right in all cases, the checks may need audit at some point (possibly needed by the merge epic). I suppose checking for EUA may not belong in this category. Ok, leave it for now, if needed we can revisit it later if auditing other checks.

File: XenAdmin/Dialogs/AcceptEuaDialog.resx [hunk 6]
```diff
@@ -0,0 +1,959 @@
+<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Dialogs/AcceptEuaDialog.resx:1]: about

> The parent should be the wizard, not the main window, otherwise it may appear hidden behind the wizard.

I don't see a mechanism to do this easily (unless I just pass the `Control` all the way from the precheck page to the `Problem`. Other checks seem to just use the mainwindow. Maybe such a mechanism could be added as part of a possible refactoring of the `Check`s?

If you want to I can pass the wizard's form all the way, but it won't look very nice :/

File: XenAdmin/Dialogs/AcceptEuaDialog.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Dialogs/AcceptEuaDialog.cs:71]: Ah true, I forgot it was a HashSet! 🤦 Still it leaves the issue of fetching 5 times the same EUA which means extra delay.

File: XenAdmin/Dialogs/AcceptEuaDialog.resx (hunk 6 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenAdmin/Dialogs/AcceptEuaDialog.resx:1]: > Other checks seem to just use the mainwindow
It depends on the case - and some cases may not even be correct (another check audit 😞 ).

In this case it is needed because if you open the EUA dialog, then click on different application on the same screen and then back to XC, the EUA dialog is hidden behind the wizard and because it is modal you can't click on the wizard, you need to click on the main window behind it to see the EUA dialog again, which is not great UX. Yes, we need to pass in the control (see for example the DMC check - when I added that I had seen an issue with adding it to the base class, I can't remember right now what exactly it was, but we can look at it again, maybe in a different PR).

**Review by CitrixChris — APPROVED** [review]

**Comment by danilo-delbusso** [comment]
Feel free to ping me on how to get and use the new RPU

**Review by CitrixChris — APPROVED** [review]

File: XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs [hunk 7]
```diff
@@ -167,6 +167,15 @@ protected override List<CheckGroup> GenerateChecks(Pool_patch patch)
             if (hotfixChecks.Count > 0)
                 groups.Add(new CheckGroup(Messages.CHECKING_UPGRADE_HOTFIX_STATUS, hotfixChecks));
 
+            if (!ManualUpgrade)
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Wizards/RollingUpgradeWizard/RollingUpgradeWizardPrecheckPage.cs:170]: nothing, I just missed it 😅

File: XenModel/Utils/Helpers.cs [hunk 8]
```diff
@@ -1580,5 +1581,33 @@ public static string AddAuthTokenToQueryString(string authToken, string existing
 
             return queryString;
         }
+
+        public static bool TryLoadHostEua(Host host, Uri targetUri, out string eua)
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenModel/Utils/Helpers.cs:1585]: Guess you're right. I had thought of it as something that didn't necessarily need to be there but we are not planning to use it anywhere else. I'll move it 👍

[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenModel/Utils/Helpers.cs:1585]: Mh actually I'd need to add a reference to JSON.net in XenAdmin for this. I'd rather not :/

File: XenAdmin/Diagnostics/Problems/UtilityProblem/EuaNotFoundProblem.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+        public sealed override string Title => string.Empty;
+
+        public override string HelpMessage => Messages.MORE_INFO;
+
+        protected override Actions.AsyncAction CreateAction(out bool cancelled)
+        {
+            Program.Invoke(Program.MainWindow, () =>
+            {
+                using (var dlg = new InformationDialog(Messages.PROBLEM_PREPARE_TO_UPGRADE))
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Diagnostics/Problems/UtilityProblem/EuaNotFoundProblem.cs:61]: ok so for this one I added a dummy action. I did think of adding a new `Problem` subclass to handle this but I'm not sure it's worth adding the extra complexity? Let me know what you think

File: XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            catch (Exception ex)
+            {
+                Log.Error(ex);
+            }
+
+            _euas = new HashSet<string>();
+            _hostsFailedToFetchEua = new HashSet<IXenObject>();
+        }
+
+        public override bool CanRun() => _targetUri != null && Hosts.Any(Helpers.Post82X);
```
[the remark this answers is not in the document]
Comment (danilo-delbusso) [reply XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs]: oh! big miss this one. well spotted! 🤦‍♂️

**Review by kc284 — APPROVED** [review]
